File: src/main/java/com/booknara/problem/array/FindAllNumbersDisappearedInAnArray.java
package com.booknara.problem.array;

import java.util.ArrayList;
import java.util.List;

/**
 * 448. Find All Numbers Disappeared in an Array (Easy)
 * https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
 */
public class FindAllNumbersDisappearedInAnArray {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }

        for (int i = 0; i < nums.length; i++) {
            int idx = Math.abs(nums[i]) - 1;
            // Keep the value negative (appearing once, twice as well)
            nums[idx] = -Math.abs(nums[idx]);
        }

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) {
                res.add(i + 1);
            }
        }

        return res;
    }
}

File: src/main/java/com/booknara/problem/array/FindAllNumbersDisappearedInAnArray.kt
package com.booknara.problem.array

/**
 * 448. Find All Numbers Disappeared in an Array (Easy)
 * https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
 */
class Solution {
    // T:O(n), S:O(1)
    fun findDisappearedNumbers(nums: IntArray): List<Int> {
        // input check, the size of array >= 1
        val res = mutableListOf<Int>()

        for ((index, value) in nums.withIndex()) {
            // println("The element at $index is $value")
            val idx = nums[Math.abs(value) - 1]
            nums[Math.abs(value) - 1] = -Math.abs(idx);
        }

        for ((index, value) in nums.withIndex()) {
            if (nums[index] > 0) {
                res.add(index + 1)
            }
        }

        return res
    }
}
